### ISSUE TITLE
Truncate commit messages to reasonable length

### ISSUE THREAD

**Issue by mike-north** [renovatebot/renovate#2893]
**What would you like Renovate to be able to do?**

When operating in conventional commits mode, I'd like the commit messages in renovate PRs to pass the [@commitlint/config-conventional](https://github.com/marionebl/commitlint/tree/master/@commitlint/config-conventional) ruleset. 

**Describe the solution you'd like**

Truncate commit message length [to 72 characters](https://github.com/marionebl/commitlint/tree/master/@commitlint/config-conventional#header-max-length).

**Describe alternatives you've considered**

* Publishing my own custom conventional commits configuration
* Manually editing Renovate PRs that violate the 72-character limit

**Additional context**

Example of an offending commit message. This can be quite easy when dealing with scoped NPM packages.

```
chore(deps): update dependency @mike-north/js-lib-renovate-config to v1.1.1
                                                                       ^
                                                                       72
```

**Comment by rarkins** [comment]
@mike-north would you be happy with simple truncation though? i.e. `chore(deps): update dependency @mike-north/js-lib-renovate-config to v1.` ?

It sounds like you should consider shortening the default commit message template too. Two suggestions:
- Remove the scope so it's just `chore: ....`
- Change [commitTopic](https://github.com/renovatebot/renovate/blob/3825c6b1cba93daff1b462cf3e4e56cd01293161/lib/config/definitions.js#L807) from the default `dependency {{depName}}` to just `{{depName}}`

The end result would be like: `chore: update @mike-north/js-lib-renovate-config to v1.1.1`

**Comment by mike-north** [comment]
Yes, the renovate commit message options would give me some more headroom for long package names.

Ideally, I'd like the package name truncated when necessary to avoid exceeding the standard 72-character maximum length in the standard commitlint settings. All other substrings are of predictable length

**Comment by rarkins** [comment]
I think tools should fit around humans and not vice versa, so I’m in general ok with these types of ideas. But on the other hand I think the standard is imposing an unreasonable limit of its not possible to include all relevant data. 

Maybe “chore: @ scope/repo v1.1.1” is even more optimised for your case?

**Comment by mike-north** [comment]
@rarkins I ended up moving some of the information out of the subject line and into the commit body

```json
"commitMessageTopic": "{{depName}}",
"commitMessage": "{{{commitMessagePrefix}}} {{{commitMessageTopic}}} {{{commitMessageSuffix}}}",
"commitMessageBody": "{{{commitMessageAction}}} {{{commitMessageExtra}}}",
```

**Comment by rarkins** [comment]
Can you give an example of the commit message/body that creates?

**Comment by jGleitz** [comment]
I just hit this as well, also because we are using @commitlint/config-conventional.

> I think tools should fit around humans and not vice versa

I agree. I would like to keep the messages because I think they are informative, but I would like to see that renovate automatically adapts them to follow conventions.

 Would you accept a PR that:

 * allows to set a maximum length for the commit message
 * truncates the commit message in a “smart” way, i.e. executing the following actions one after the other until the message is short enough:
     * set `commitMessageTopic` to `{depName}` if it is set to the default
     * remove the dependency group (or anything before a `/` or `:`) from `{depName}`
     * set `commitMessageTopic` to `null`
     * set `commitMessageExtra` to `null`
     *  truncate the message

If the message was altered in any way, `commitMessageBody` would be prepended with the original message.

**Comment by rarkins** [comment]
@jGleitz sorry for the delay in getting back to you. I've been thinking about it and will respond soon

**Comment by jGleitz** [comment]
@rarkins: any updates?

**Comment by rarkins** [comment]
Some questions:

- Would you perform these steps in order, and stop once you have reached the desired length? Or perform all of them if the original length is too long?
- Would the PR title remain the same, or you would truncate that too? Currently it's copied from the commitMessage in most cases

**Comment by jGleitz** [comment]
> Would you perform these steps in order, and stop once you have reached the desired length? Or perform all of them if the original length is too long?

I would perform them in order and stop as soon as we reached the desired length.

> Would the PR title remain the same, or you would truncate that too? Currently it's copied from the commitMessage in most cases

I, personally, would not change the PR title but leave it in its long version. However, I have no strong opinion on this. Maybe it might be desirable to have a shorter PR title as well, or to have the title always match the commit message.

**Comment by rarkins** [comment]
OK. Proposed new config option: `commitMessageMaximumLength`.

If set, use the logic that you described.

Here's where the `prTitle` is compiled, or defaults to the `commitMessage` value: https://github.com/renovatebot/renovate/blob/4104ba04e3c58063163bc0ae7091fdc2be2d6f97/lib/workers/repository/updates/generate.ts#L217-L229

`prTitle` should still default to using the full `commitMessage`.

**Comment by ViennaL** [comment]
@rarkins and @jGleitz, Any chance we get this smart commit message truncation into a release in the near future?
Currently, we cannot update some dependencies in our projects without breaking our commit linting jobs.
I would really appreciate any feedback - thanks in advance.

**Comment by viceice** [comment]
This needs a community contribution. It seems @jGleitz is no longer working on this.

**Comment by rarkins** [comment]
There is a `replace` directive available for handlebars which you could possibly use to reduce any particularly long strings.

**Comment by ViennaL** [comment]
> There is a `replace` directive available for handlebars which you could possibly use to reduce any particularly long strings.

@rarkins , Can you please direct me to an example or some documentation about this feature? Thanks in advance.
